source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,210 @@
+    if (context.hasResponded(this)) {
+      return;
+    }
+    if (context.incrementAndGetCount(this) == verifiers_.size() || Status::Ok == status) {
+      context.responded(this);
+      getCallback(context)->onComplete(status, context);
+    }
+  }
+};
+
+// Requires all verifier
+class AllVerifierImpl : public BaseGroupVerifierImpl, public VerifierCallbacks {
+public:
+  AllVerifierImpl(const JwtRequirementAndList& and_list, const AuthFactory& factory,
+                  const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                  const Extractor& extractor) {
+    for (const auto& it : and_list.requirements()) {
+      auto verifier = Verifier::create(it, providers, factory, extractor);
+      verifier->registerCallback(this);
+      verifiers_.push_back(std::move(verifier));
+    }
+  }
+
+  void onComplete(const Status& status, VerifyContext& context) override {
+    if (context.hasResponded(this)) {
+      return;
+    }
+    if (context.incrementAndGetCount(this) == verifiers_.size() || Status::Ok != status) {
+      context.responded(this);
+      getCallback(context)->onComplete(status, context);
+    }
+  }
+};
+
+// Match all, for requirement not set
+class AllowAllVerifierImpl : public BaseVerifierImpl {
+public:
+  void verify(VerifyContext& context) override {
+    getCallback(context)->onComplete(Status::Ok, context);
+  }
+};
+
+} // namespace
+
+VerifierPtr Verifier::create(const JwtRequirement& requirement,
+                             const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                             const AuthFactory& factory, const Extractor& extractor) {
+  std::string provider_name;
+  std::vector<std::string> audiences;
+  switch (requirement.requires_type_case()) {
+  case JwtRequirement::RequiresTypeCase::kProviderName:
+    provider_name = requirement.provider_name();
+    break;
+  case JwtRequirement::RequiresTypeCase::kProviderAndAudiences:
+    for (const auto& it : requirement.provider_and_audiences().audiences()) {
+      audiences.emplace_back(it);
+    }
+    provider_name = requirement.provider_and_audiences().provider_name();
+    break;
+  case JwtRequirement::RequiresTypeCase::kRequiresAny:
+    return std::make_unique<AnyVerifierImpl>(requirement.requires_any(), factory, providers,
+                                             extractor);
+  case JwtRequirement::RequiresTypeCase::kRequiresAll:
+    return std::make_unique<AllVerifierImpl>(requirement.requires_all(), factory, providers,
+                                             extractor);
+  case JwtRequirement::RequiresTypeCase::kAllowMissingOrFailed:
+    return std::make_unique<AllowFailedVerifierImpl>(factory, extractor,
+                                                     requirement.allow_missing_or_failed().value());
+  case JwtRequirement::RequiresTypeCase::REQUIRES_TYPE_NOT_SET:
+    return std::make_unique<AllowAllVerifierImpl>();
```

Comment:
I don't think we need to handle NOT_SET case.  Just assert it with NOT_REACHED_.

Reply:
A rule can have a match without requires. In that case the request is always allowed. Falling through to NOT_REACHED_ panic is not the correct behavior.